SIL.Core/Xml/XmlUtils.cs

```diff
@@ -762,5 +1191,159 @@ private static bool IsLegalXmlChar(int codePoint)
+		{
+			bool fSuccessfulVisit = false;
+			if (input.HasAttributes) // can be, e.g, if Input is a XmlTextNode
+			{
+				foreach (XAttribute xa in input.Attributes())
+				{
+					if (visitor.Visit(xa))
+						fSuccessfulVisit = true;
+				}
+			}
+			if (input.Elements().Any()) // not sure whether this can happen.
+			{
+				foreach (var child in input.Elements())
+				{
+					if (VisitAttributes(child, visitor))
+						fSuccessfulVisit = true;
+				}
+			}
+			return fSuccessfulVisit;
+		}
+
+		#region Serialize/Deserialize
+
+		/// <summary>
+		/// Try to serialize the given object into an xml string
+		/// </summary>
+		/// <param name="objToSerialize"></param>
+		/// <returns>empty string if couldn't serialize object</returns>
+		public static string SerializeObjectToXmlString(object objToSerialize)
+		{
+			string settingsXml = "";
+			using (MemoryStream stream = new MemoryStream())
+			{
+				using (XmlTextWriter textWriter = new XmlTextWriter(stream, Encoding.UTF8))
+				{
+					XmlSerializer xmlSerializer = new XmlSerializer(objToSerialize.GetType());
+					xmlSerializer.Serialize(textWriter, objToSerialize);
+					textWriter.Flush();
+					stream.Seek(0, SeekOrigin.Begin);
+					XmlDocument doc = new XmlDocument();
+					doc.Load(stream);
+					settingsXml = doc.OuterXml;
+				}
+			}
+			return settingsXml;
+		}
+
+		/// <summary>
+		/// Deserialize the given xml string into an object of targetType class
+		/// </summary>
+		/// <param name="xml"></param>
+		/// <param name="targetType"></param>
+		/// <returns>null if we didn't deserialize the object</returns>
+		public static object DeserializeXmlString(string xml, Type targetType)
+		{
+			// TODO-Linux: System.Boolean System.Type::op_{Ine,E}quality(System.Type,System.Type)
+			// is marked with [MonoTODO] and might not work as expected in 4.0.
+			if (String.IsNullOrEmpty(xml) || targetType == null)
+				return null;
+			using (MemoryStream stream = new MemoryStream())
+			{
+				using (XmlTextWriter textWriter = new XmlTextWriter(stream, Encoding.UTF8))
+				{
+					XmlDocument doc = new XmlDocument();
+					doc.LoadXml(xml);
+					// get the type from the xml itself.
+					// if we can find an existing class/type, we can try to deserialize to it.
+					if (targetType != null)
+					{
+						doc.WriteContentTo(textWriter);
+						textWriter.Flush();
+						stream.Seek(0, SeekOrigin.Begin);
+						XmlSerializer xmlSerializer = new XmlSerializer(targetType);
+						try
+						{
+							object settings = xmlSerializer.Deserialize(stream);
+							return settings;
+						}
+						catch
+						{
+							// something went wrong trying to deserialize the xml
+							// perhaps the structure of the stored data no longer matches the class
+						}
+					}
+				}
+			}
+			return null;
+		}
+
+		#endregion Serialize/Deserialize
+	}
+
+	/// <summary>
+	/// Interface for operations we can apply to attributes.
+	/// </summary>
+	public interface IAttributeVisitor
```

Comment:
If ReplaceSubstringInAttr is the only IAttributeVisitor then this seems overengineered. VisitAttributes could be replaced by a ReplaceSubstringInAttributes method.